### PR TITLE
Pass through package analysis warnings to analysis report

### PR DESCRIPTION
**Pull request by sunandabalu**
This change is to pass through the useful warnings in the `TargetCompatibilityReferenceAnalyzer` to the AnalysisReport

### CODE REVIEW

File: src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/NuGetReference.cs [hunk 1]
```diff
@@ -18,5 +19,7 @@ public override string ToString()
         }
 
         public string? PrivateAssets { get; set; }
+
+        public IEnumerable<string>? ActionDetails { get; set; }
```
Comment (twsouthwick) [inline src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/NuGetReference.cs:23]: Can you make this non-nullable? Is there a reason you want this nullable?

Comment (twsouthwick) [reply src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/NuGetReference.cs:23]: Can you make this (and the other properties while you're at it) `get; init;` instead of `set;`?

File: src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/AnalyzePackageStatus.cs [hunk 2]
```diff
@@ -101,7 +101,15 @@ void GetResults<T>(string name, string action, IReadOnlyCollection<T> collection
                     {
                         foreach (var s in collection)
                         {
-                            results.Add(string.Concat(name, s, action));
+                            var mc = s as NuGetReference;
```
Comment (twsouthwick) [inline src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/AnalyzePackageStatus.cs:104]: Why the type check here? If you do need it, better to switch to pattern matching

File: src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/NuGetReference.cs (hunk 1 above)
Comment (twsouthwick) [reply src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/NuGetReference.cs:23]: I'm also confused as to why this is here. Why is this on this object? I'm not understanding why a NuGetReference would need to have an action details

**Review by twsouthwick — CHANGES_REQUESTED** [review]

File: src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/NuGetReference.cs [hunk 3]
```diff
@@ -2,6 +2,7 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System;
+using System.Collections.Generic;
```
Comment (twsouthwick) [inline src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/NuGetReference.cs:5]: nit: is this necessary?

File: src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/Dependencies/IDependencyCollection.cs [hunk 4]
```diff
@@ -1,18 +1,40 @@
...
+using System;
 using System.Collections.Generic;
+using System.Linq;
 
 namespace Microsoft.DotNet.UpgradeAssistant.Dependencies
 {
     public interface IDependencyCollection<T> : IEnumerable<T>
     {
-        bool Add(T item, BuildBreakRisk risk = BuildBreakRisk.None);
+        bool Add(T item, OperationDetails details);
```
Comment (twsouthwick) [inline src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/Dependencies/IDependencyCollection.cs:12]: So this will break current implements (ie loose dependency). Can we phase in the new API?

what about 

```csharp
[Obsolete]
[EditorBrowseable(EditorBrowseableState.None)]
bool Add(T item, BuildBreakRisk risk = BuildBreakRisk.None);

bool Add(T item, OperationDetails detail);

...etc
```

File: src/extensions/vb/Microsoft.DotNet.UpgradeAssistant.Extensions.VisualBasic/MyDotAnalyzer.cs [hunk 5]
```diff
@@ -70,7 +70,7 @@ public async Task AnalyzeAsync(IProject project, IDependencyAnalysisState state,
                     // 1>C:\{ProjectDir}\My Project\Settings.Designer.vb(57,10): error BC30002: Type 'Global.System.Configuration.UserScopedSettingAttribute' is not defined.
                     // 1>C:\{ProjectDir}\My Project\Settings.Designer.vb(59,10): error BC30002: Type 'Global.System.Configuration.DefaultSettingValueAttribute' is not defined.
                     _logger.LogInformation("Reference to configuration package ({SystemConfigurationPackageName}, version {SystemConfigurationPackageVersion}) needs added", SystemConfigurationPackageName, systemConfigurationPackage.Version);
-                    state.Packages.Add(systemConfigurationPackage);
+                    state.Packages.Add(systemConfigurationPackage, new());
```
Comment (twsouthwick) [inline src/extensions/vb/Microsoft.DotNet.UpgradeAssistant.Extensions.VisualBasic/MyDotAnalyzer.cs:73]: You can remove this with the support kept for BuildBreakRisk

File: src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/Dependencies/IDependencyCollection.cs [hunk 6]
```diff
@@ -1,18 +1,40 @@
...
+    {
+        public Operation(T item, OperationDetails details)
+        {
+            Item = item;
+            Action = details;
+        }
+
+        public T Item { get; }
+
+        public OperationDetails Action { get; }
```
Comment (twsouthwick) [inline src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/Dependencies/IDependencyCollection.cs:38]: Action doesn't fit. What about Details?

File: src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/DependencyCollection{T}.cs [hunk 7]
```diff
@@ -62,22 +76,32 @@ public IEnumerator<T> GetEnumerator()
...
-                yield return additions;
+                yield return additions.Item;
             }
         }
 
         IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
 
+        public bool Add(T item, OperationDetails details)
+        {
+            throw new NotImplementedException();
```
Comment (twsouthwick) [inline src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/DependencyCollection{T}.cs:93]: ?

File: tests/tool/Integration.Tests/IntegrationScenarios/PCL/Upgraded/SamplePCL.csproj [hunk 8]
```diff
@@ -4,15 +4,15 @@
...
-    <Error Condition="Exists('..\packages\Microsoft.Bcl.Build.1.0.14\tools\Microsoft.Bcl.Build.targets')" Text="The build restored NuGet packages. Build the project again to include these packages in the build. For more information, see http://go.microsoft.com/fwlink/?LinkID=317568." HelpKeyword="BCLBUILD2002" />
-  </Target>
   <ItemGroup>
-    <PackageReference Include="Newtonsoft.Json" Version="9.0.1" />
     <PackageReference Include="Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers" Version="0.2.222702">
       <PrivateAssets>all</PrivateAssets>
     </PackageReference>
+    <PackageReference Include="Newtonsoft.Json" Version="9.0.1" />
   </ItemGroup>
+  <Import Project="..\packages\Microsoft.Bcl.Build.1.0.14\tools\Microsoft.Bcl.Build.targets" Condition="Exists('..\packages\Microsoft.Bcl.Build.1.0.14\tools\Microsoft.Bcl.Build.targets')" />
```
Comment (twsouthwick) [inline tests/tool/Integration.Tests/IntegrationScenarios/PCL/Upgraded/SamplePCL.csproj:13]: Why this change?

**Review by twsouthwick — CHANGES_REQUESTED** [review]

File: src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/Dependencies/IDependencyCollection.cs (hunk 6 above)
Comment (sunandabalu) [reply src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/Dependencies/IDependencyCollection.cs:38]: The member in the OperationDetails is also called details, so when I was accessing it - it was `Details.details` which looks super weird.

File: tests/tool/Integration.Tests/IntegrationScenarios/PCL/Upgraded/SamplePCL.csproj (hunk 8 above)
Comment (sunandabalu) [reply tests/tool/Integration.Tests/IntegrationScenarios/PCL/Upgraded/SamplePCL.csproj:13]: The test changed the ordered in which the csproj is constructed so had to make this change to unblock tests in CI too.

File: src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/DependencyCollection{T}.cs (hunk 7 above)
Comment (sunandabalu) [reply src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/DependencyCollection{T}.cs]: removed

File: src/extensions/vb/Microsoft.DotNet.UpgradeAssistant.Extensions.VisualBasic/MyDotAnalyzer.cs (hunk 5 above)
Comment (sunandabalu) [reply src/extensions/vb/Microsoft.DotNet.UpgradeAssistant.Extensions.VisualBasic/MyDotAnalyzer.cs]: removing throws a warning, so made the call explicit.

File: src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/Dependencies/IDependencyCollection.cs [hunk 9]
```diff
@@ -1,18 +1,49 @@
...
 
 namespace Microsoft.DotNet.UpgradeAssistant.Dependencies
 {
     public interface IDependencyCollection<T> : IEnumerable<T>
     {
+        bool Add(T item, OperationDetails details);
+
+        bool Remove(T item, OperationDetails details);
+
+        [Obsolete("Use the newer Add API", error: false)]
```
Comment (twsouthwick) [inline src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/Dependencies/IDependencyCollection.cs:17]: nit: I'd explain a bit more, such as "This API will be removed sometime after September 1, 2021. Please refactor to use a supported overload"

File: src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/Dependencies/IDependencyCollection.cs [hunk 10]
```diff
@@ -1,18 +1,49 @@
...
+        [EditorBrowsable(EditorBrowsableState.Never)]
         bool Add(T item, BuildBreakRisk risk = BuildBreakRisk.None);
 
+        [Obsolete("Use the newer Remove API", error: false)]
+        [EditorBrowsable(EditorBrowsableState.Never)]
         bool Remove(T item, BuildBreakRisk risk = BuildBreakRisk.None);
 
-        IReadOnlyCollection<T> Additions { get; }
+        IReadOnlyCollection<Operation<T>> Additions { get; }
```
Comment (twsouthwick) [inline src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/Dependencies/IDependencyCollection.cs:25]: What about the old interface properties of Additions, and Removals? Not sure if that can be shimmed over, just wanted to know it was intentional.

File: src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/Dependencies/IDependencyCollection.cs (hunk 6 above)
Comment (twsouthwick) [reply src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/Dependencies/IDependencyCollection.cs:47]: Ok. But what does action mean here, then?

File: src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/AnalyzePackageStatus.cs [hunk 11]
```diff
@@ -95,13 +95,20 @@ private static IReadOnlyCollection<string> ExtractAnalysisResult(IDependencyAnal
...
 
-                void GetResults<T>(string name, string action, IReadOnlyCollection<T> collection)
+                void GetResults<T>(string name, string action, IReadOnlyCollection<Operation<T>> collection)
                 {
                     if (collection.Any())
                     {
                         foreach (var s in collection)
                         {
-                            results.Add(string.Concat(name, s, action));
+                            if (s.Action != null && s.Action.Details != null && s.Action.Details.Any() && !results.Intersect(s.Action.Details).Any())
```
Comment (twsouthwick) [inline src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/AnalyzePackageStatus.cs:104]: Why so many null checks? Are these non-nullable?

Looks like you want to merge these, right? why not make the collection a hashset that will handle duplicates?

File: src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/Analyzers/TargetCompatibilityReferenceAnalyzer.cs [hunk 12]
```diff
@@ -65,16 +68,23 @@ public async Task AnalyzeAsync(IProject project, IDependencyAnalysisState state,
 
                         if (isMajorChange)
                         {
+                            details.Add(string.Concat("Package ", packageReference.Name, " needs to be upgraded across major versions (", packageReference.Version, " -> ", updatedReference.Version, ") which may introduce breaking changes"));
                             _logger.LogWarning("Package {NuGetPackage} has been upgraded across major versions ({OldVersion} -> {NewVersion}) which may introduce breaking changes", packageReference.Name, packageReference.Version, updatedReference.Version);
                         }
 
                         if (updatedReference.IsPrerelease)
                         {
+                            details.Add(string.Concat("Package ", packageReference.Name, " needs to be upgraded to a prerelease version (", updatedReference.Version, ") because no released version supports target(s) ", string.Join(", ", state.TargetFrameworks)));
```
Comment (twsouthwick) [inline src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/Analyzers/TargetCompatibilityReferenceAnalyzer.cs:77]: This is a bunch of string concatenation going on here. Can you swap it out to use string.Format (or better yet, SR.Format)? Maybe also consider using resource files

File: src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/DependencyCollection{T}.cs [hunk 13]
```diff
@@ -62,22 +77,60 @@ public IEnumerator<T> GetEnumerator()
...
         }
 
         IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
 
+        [Obsolete("Use the newer Add API", error: false)]
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        bool IDependencyCollection<T>.Add(T item, BuildBreakRisk risk)
+        {
+            var operation = new Operation<T>(item, new() { Risk = risk });
+            if (Contains(item))
```
Comment (twsouthwick) [inline src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/DependencyCollection{T}.cs:97]: This is duplicating already existing logic. Can't you forward on the call by `Add(item, new() {Risk = risk})`?

File: tests/tool/Integration.Tests/IntegrationScenarios/PCL/Upgraded/SamplePCL.csproj (hunk 8 above)
Comment (twsouthwick) [reply tests/tool/Integration.Tests/IntegrationScenarios/PCL/Upgraded/SamplePCL.csproj:13]: What test change did that? I don't see anything that should be changing the order

**Review by twsouthwick — CHANGES_REQUESTED** [review]

File: src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/DependencyCollection{T}.cs (hunk 13 above)
Comment (sunandabalu) [reply src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/DependencyCollection{T}.cs:97]: I tried that, it was not able to resolve the other Add method.

File: src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/Dependencies/IDependencyCollection.cs (hunk 10 above)
Comment (sunandabalu) [reply src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/Dependencies/IDependencyCollection.cs:25]: its going to be hard to shim the Additions and Deletions collection.

File: src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/AnalyzePackageStatus.cs (hunk 11 above)
Comment (sunandabalu) [reply src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/AnalyzePackageStatus.cs:104]: yes they are all non-nullable.
